js/spreadsheet.js

```diff
@@ -67,248 +67,131 @@ var spreadsheet = function(options) {
-      firstCol = startAt[1];
-
-      for (var i = firstCol + 1; i < allData.length; i++) {
-        if (allData[startAt[0]][i] === null) {
-          lastCol = i - 1;
-          break;
-        }
-      }
+  function getDataCoordinates(tableData, cellPosition, processedCells) {
+    if (!processedCells) {
+      processedCells = [];
+      processedCells[cellPosition.x] = [];
+      processedCells[cellPosition.x][cellPosition.y] = true;
+    }
 
-      for (var i = startAt[0]; i > 0; i--) {
-        if (allData[i][lastCol] === null) {
-          firstRow = i ? i - 1 : i;
-        }
-      }
+    var cellsGroup = getNearestCells(cellPosition);
 
-      firstRow = firstRow || 0;
+    cellsGroup['currentCell'] = cellPosition;
 
-      for (var i = firstRow; i < allData.length; i++) {
-        if (allData[i][lastCol] === null) {
-          lastRow = i - 1;
-          break;
-        }
-      }
-    } else if (moveTo.top) {
-      // When data located on the top of the selected cell
-      lastRow = startAt[0];
+    for (var cell in cellsGroup) {
+      var hasData = tableData[cellsGroup[cell].x][cellsGroup[cell].y] !== null;
+      var processedRow = processedCells[cellsGroup[cell].x];
 
-      for (var i = lastRow - 1; i > 0; i--) {
-        if (allData[i][startAt[1]] === null) {
-          firstRow = i;
-          break;
-        }
+      if (!processedRow) {
+        processedCells[cellsGroup[cell].x] = [];
       }
 
-      firstRow = firstRow || 0;
+      var processedCell = processedCells[cellsGroup[cell].x][cellsGroup[cell].y];
 
-      for (var i = startAt[1]; i > 0; i--) {
-        if (allData[firstRow][i] === null) {
-          firstCol = i ? i + 1 : i;
-          break;
-        }
+      if (!processedCell) {
+        processedCells[cellsGroup[cell].x][cellsGroup[cell].y] = false;
       }
 
-      firstCol = firstCol || 0;
-
-      for (var i = firstCol; i < allData.length; i++) {
-        if (allData[firstRow][i] === null) {
-          lastCol = i - 1;
-          break;
-        }
-      }
-    } else if (moveTo.bottom) {
-      // When data located on the bottom of the selected cell
-      firstRow = startAt[0];
+      if (hasData && (!processedRow || !processedCell)) {
+        processedCells[cellsGroup[cell].x][cellsGroup[cell].y] = true;
+        getDataCoordinates(tableData, cellsGroup[cell], processedCells);
+      }      
+    }
 
-      for (var i = firstRow + 1; i < allData.length; i++) {
-        if (allData[i][startAt[1]] === null) {
-          lastRow = i - 1;
-          break;
-        }
-      }
+    return processedCells;
+  }
 
-      for (var i = startAt[1]; i > 0; i--) {
-        if (allData[lastRow][i] === null) {
-          firstCol = i + 1;
-          break;
+  /**
+   * The method that returns us a range that we should select in the table
+   * 
+   * @param {array} processedCells array processed by getDataCoordinates function
+   */
+  function getSelectionCoordinates(processedCells) {
+    var rowIndexes = [];
```

Comment:
The plural for **index** is **indices**.

```suggestion
    var rowIndices = [];
```